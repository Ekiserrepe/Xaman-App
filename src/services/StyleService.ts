/**
 * Style service
 */
import { has, get } from 'lodash';
import { Appearance, StyleSheet } from 'react-native';

import { Images } from '@common/helpers/images';

import CoreRepository from '@store/repositories/core';
import CoreModel from '@store/models/objects/core';

import { Themes } from '@store/types';

import { ColorsGeneral, ColorsTheme } from '@theme/colors';

import { ImageStyle, TextStyle, ViewStyle } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';
/* Types  ==================================================================== */
type NamedStyles<T> = { [P in keyof T]: ViewStyle | TextStyle | ImageStyle };
export type StyleType = Record<string, any>;

/* Service  ==================================================================== */
class StyleService {
    private themeName: Themes;
    private currentStyle: any;

    public hairlineWidth = StyleSheet.hairlineWidth;

    constructor() {
        this.themeName = 'light';
        this.currentStyle = { ...ColorsTheme.light, ...ColorsGeneral };
    }

    initialize = (coreSettings: CoreModel) => {
        return new Promise<void>((resolve, reject) => {
            try {
                // default theme
                let theme = 'light' as Themes;

                // if app is not initialized set default theme from OS settings
                if (!coreSettings?.initialized) {
                    // try to get user's appearance preferences from OS
                    const colorScheme = Appearance.getColorScheme();
                    if (colorScheme) {
                        theme = colorScheme;

                        // persist the changes
                        CoreRepository.saveSettings({
                            theme: colorScheme,
                        });
                    }
                } else {
                    // lowerCase the theme name is required as we stored
                    // theme name in title mode in version before v1.0.1
                    theme = (coreSettings.theme?.toLowerCase() || 'light') as Themes;
                }

                this.setTheme(theme);

                resolve();
            } catch (e) {
                reject(e);
            }
        });
    };

    setTheme = (theme: Themes) => {
        if (this.themeName !== theme && has(ColorsTheme, theme)) {
            this.themeName = theme;
            this.currentStyle = { ...ColorsTheme[theme], ...ColorsGeneral };
        }
    };

    create = <T extends NamedStyles<T> | NamedStyles<any>>(styles: T | NamedStyles<T>): T => {
        return Object.entries(styles).reduce((themed: any, style: any) => {
            const [key, value] = style;

            return { ...themed, [key]: this.applyTheme(value) };
        }, {});
    };

    applyTheme = (style: StyleType) => {
        if (typeof style === 'object') {
            return Object.entries(style).reduce((themed: any, item: any) => {
                const [key, value] = item;

                return { ...themed, [key]: this.value(value) };
            }, {});
        }
        return style;
    };

    value = (value: string): string => {
        if (this.isReference(value)) {
            const referenceKey: string = this.createKeyFromReference(value);
            return this.findValue(referenceKey);
        }
        return value;
    };

    private findValue = (name: string): any => {
        return this.currentStyle[name];
    };

    private isReference = (value: string): boolean => {
        return `${value}`.startsWith('$');
    };

    private createKeyFromReference = (value: string): string => {
        return `${value}`.substring(1);
    };

    isDarkMode = (): boolean => {
        return this.themeName !== 'light';
    };

    getBackdropInterpolateColor = () => {
        if (this.themeName === 'dark' || this.themeName === 'moonlight') {
            return ['rgba(50,50,50,0)', 'rgba(50,50,50,0.7)'];
        }

        return ['rgba(0,0,0,0)', 'rgba(0,0,0,0.8)'];
    };

    select<T extends string | number>(spec: { light?: T; dark?: T; default?: T }): T | undefined {
        return 'light' in spec && !this.isDarkMode()
            ? spec.light
            : 'dark' in spec && this.isDarkMode()
              ? spec.dark
              : spec.default;
    }

    getImage = (image: Extract<keyof typeof Images, string>): { uri: string } => {
        // if dark mode and there is a light mode for image return light
        if (this.isDarkMode() && `${image}Light` in Images) {
            return get(Images, `${image}Light`)!;
        }

        return get(Images, image);
    };

    getImageIfLightModeIfDarkMode = (
        imageIfLightMode: Extract<keyof typeof Images, string>,
        imageIfDarkMode: Extract<keyof typeof Images, string>,
    ): { uri: string } => {
        // if dark mode and there is a light mode for image return light
        if (this.isDarkMode() && `${imageIfDarkMode}` in Images) {
            return get(Images, `${imageIfDarkMode}`)!;
        }

        return get(Images, imageIfLightMode);
    };

    getCurrentTheme = (): string => {
        return this.themeName;
    };
}

export default new StyleService();
