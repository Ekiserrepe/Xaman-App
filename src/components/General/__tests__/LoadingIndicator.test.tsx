import React from 'react';
import renderer from 'react-test-renderer';

import StyleService from '@services/StyleService';

import { LoadingIndicator } from '../LoadingIndicator';

describe('[LoadingIndicator]', () => {
    it('renders with default props', () => {
        const tree = renderer.create(<LoadingIndicator />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders with specified color (light)', () => {
        const spyStyleService = jest.spyOn(StyleService, 'value');
        const tree = renderer.create(<LoadingIndicator color="light" />).toJSON();
        expect(tree).toMatchSnapshot();
        expect(spyStyleService).toHaveBeenCalledWith('$white');
    });

    it('renders with specified color (dark)', () => {
        const spyStyleService = jest.spyOn(StyleService, 'value');
        const tree = renderer.create(<LoadingIndicator color="dark" />).toJSON();
        expect(tree).toMatchSnapshot();
        expect(spyStyleService).toHaveBeenCalledWith('$black');
    });

    it('renders with specified size (large)', () => {
        const tree = renderer.create(<LoadingIndicator size="large" />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders while animating is false', () => {
        const tree = renderer.create(<LoadingIndicator animating={false} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders with custom styles', () => {
        const customStyle = { margin: 10 };
        const tree = renderer.create(<LoadingIndicator style={customStyle} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
