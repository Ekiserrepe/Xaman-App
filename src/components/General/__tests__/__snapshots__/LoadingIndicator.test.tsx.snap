// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[LoadingIndicator] renders while animating is false 1`] = `
<ActivityIndicator
  animating={false}
  color="#000000"
  size="small"
/>
`;

exports[`[LoadingIndicator] renders with custom styles 1`] = `
<ActivityIndicator
  animating={true}
  color="#000000"
  size="small"
  style={
    {
      "margin": 10,
    }
  }
/>
`;

exports[`[LoadingIndicator] renders with default props 1`] = `
<ActivityIndicator
  animating={true}
  color="#000000"
  size="small"
/>
`;

exports[`[LoadingIndicator] renders with specified color (dark) 1`] = `
<ActivityIndicator
  animating={true}
  color="#000000"
  size="small"
/>
`;

exports[`[LoadingIndicator] renders with specified color (light) 1`] = `
<ActivityIndicator
  animating={true}
  color="#ffffff"
  size="small"
/>
`;

exports[`[LoadingIndicator] renders with specified size (large) 1`] = `
<ActivityIndicator
  animating={true}
  color="#000000"
  size="large"
/>
`;
